src/main/java/io/temporal/internal/replay/ReplayWorkflowContextImpl.java

```diff
@@ -214,183 +207,127 @@ public SearchAttributes getSearchAttributes() {
   }
 
   @Override
-  public Consumer<Exception> scheduleActivityTask(
-      ExecuteActivityParameters parameters, BiConsumer<Optional<Payloads>, Exception> callback) {
-    return activityClient.scheduleActivityTask(parameters, callback);
+  public Functions.Proc1<Exception> scheduleActivityTask(
+      ExecuteActivityParameters parameters, Functions.Proc2<Optional<Payloads>, Failure> callback) {
+    ScheduleActivityTaskCommandAttributes.Builder attributes = parameters.getAttributes();
+    if (attributes.getActivityId().isEmpty()) {
+      attributes.setActivityId(workflowStateMachines.randomUUID().toString());
+    }
+    Functions.Proc cancellationHandler =
+        workflowStateMachines.scheduleActivityTask(parameters, callback);
+    return (exception) -> cancellationHandler.apply();
   }
 
   @Override
-  public Consumer<Exception> scheduleLocalActivityTask(
+  public Functions.Proc scheduleLocalActivityTask(
       ExecuteLocalActivityParameters parameters,
-      BiConsumer<Optional<Payloads>, Exception> callback) {
-    return workflowClock.scheduleLocalActivityTask(parameters, callback);
+      Functions.Proc2<Optional<Payloads>, Failure> callback) {
+    return workflowStateMachines.scheduleLocalActivityTask(parameters, callback);
   }
 
   @Override
-  public Consumer<Exception> startChildWorkflow(
+  public Functions.Proc1<Exception> startChildWorkflow(
       StartChildWorkflowExecutionParameters parameters,
-      Consumer<WorkflowExecution> executionCallback,
-      BiConsumer<Optional<Payloads>, Exception> callback) {
-    return workflowClient.startChildWorkflow(parameters, executionCallback, callback);
+      Functions.Proc1<WorkflowExecution> executionCallback,
+      Functions.Proc2<Optional<Payloads>, Exception> callback) {
+    Functions.Proc cancellationHandler =
+        workflowStateMachines.startChildWorkflow(parameters, executionCallback, callback);
+    return (exception) -> cancellationHandler.apply();
   }
 
   @Override
-  public Consumer<Exception> signalWorkflowExecution(
+  public Functions.Proc1<Exception> signalExternalWorkflowExecution(
```

Comment:
Interesting that signalExternalWorkflowExecution is cancelable. What is the scenario for that?

Reply:
When workflow (or any part of its code) is canceled we automatically cancel all outstanding requests including signalExternalWorkflowExecution. Note that it is cancellable only until the command was sent out to the service.

Reply:
yeah, that cancellation window seems very, very tiny tbh.